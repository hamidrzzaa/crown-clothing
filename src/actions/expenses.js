//action creators for Expense
import uuid from "uuid";

export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    payload: { id: uuid.v4(), description, note, amount, createdAt }
  };
};

export const removeExpense = id => ({ type: "REMOVE_EXPENSE", payload: id });

export const editExpense = (id, updates) => {
  return {
    type: "EDIT_EXPENSE",
    payload: { id, updates }
  };
};
