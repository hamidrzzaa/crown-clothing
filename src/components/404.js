import React from "react";
import { Link } from "react-router-dom";

export const PageNotFound = () => (
  <div>
    <h1>
      404! -<Link to="/">Go Home</Link>
    </h1>
  </div>
);

export default PageNotFound;
