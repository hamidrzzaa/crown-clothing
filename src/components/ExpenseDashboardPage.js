import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilters";

export const ExpenseDashboardPage = () => {
  return (
    <div>
      <ExpenseListFilter />
      <ExpenseListFilter />
    </div>
  );
};

export default ExpenseDashboardPage;
