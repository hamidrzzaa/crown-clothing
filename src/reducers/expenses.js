const expensesReducerInitialState = [];

const expensesReducer = (state = expensesReducerInitialState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.payload];
    case "REMOVE_EXPENSE":
      const newState = state.filter(expense => expense.id !== action.payload);
      return newState;
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.payload.id) {
          return { ...expense, ...action.payload.updates };
        } else {
          return expense;
        }
      });

    default:
      return state;
  }
};

export default expensesReducer;
