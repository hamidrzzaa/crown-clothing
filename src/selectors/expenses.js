import moment from "moment";
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      let startDateMatch = startDate
        ? startDate.isSameOrBefore(moment(expense.createdAt), "day")
        : true;
      let endDateMatch = endDate
        ? endDate.isSameOrAfter(moment(expense.createdAt), "day")
        : true;
      let textMatch =
        expense.description.toLowerCase().includes(text) ||
        expense.note.toLowerCase().includes(text);
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((expA, expB) => {
      if (sortBy === "date") {
        return expA.createdAt > expB.createdAt ? -1 : +1;
      }
      if (sortBy === "amount") {
        return expA.amount <= expB.amount ? 1 : -1;
      }
    });
};
