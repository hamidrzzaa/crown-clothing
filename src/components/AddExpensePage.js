import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export const AddExpensePage = props => {
  const onSubmit = data => {
    props.addExpense(data);
    props.history.push("/");
  };
  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={onSubmit} />
    </div>
  );
};

// const mapDispatchToProps = dispatch => ({
//   addExpense: data => dispatch(addExpense(data))
// });

export default connect(null, { addExpense })(AddExpensePage);
