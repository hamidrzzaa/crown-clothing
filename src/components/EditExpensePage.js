import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";
export const EditExpensePage = props => {
  const onFromSubmit = date => {
    props.editExpense(props.expense.id, date);
    props.history.push("/");
  };

  const onExpenseDelete = () => {
    props.removeExpense(props.expense.id);
    props.history.push("/");
  };

  if (props.expense) {
    return (
      <div>
        <ExpenseForm onSubmit={onFromSubmit} expense={props.expense} />
        <button onClick={onExpenseDelete}>Remove</button>
      </div>
    );
  } else {
    return (
      <div>
        <p>There is no expense </p>
      </div>
    );
  }
};

const mapStateToProps = (state, ownProps) => {
  const idParam = ownProps.match.params.id;
  const filter = state.expenses.filter(exp => exp.id === idParam);

  return {
    expense: filter[0]
  };
};

export default connect(mapStateToProps, { editExpense, removeExpense })(
  EditExpensePage
);
