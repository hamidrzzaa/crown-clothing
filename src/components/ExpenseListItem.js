import React from "react";
import { removeExpense } from "../actions/expenses";
import { Link } from "react-router-dom";
export const ExpenseListItem = props => {
  const date = new Date(props.createdAt);

  return (
    <div>
      <h3>{props.description}</h3>
      <p>{props.amount}</p>
      <p>
        {date.getHours()}:{date.getMinutes()}
      </p>

      <button>
        {" "}
        <Link to={`/edit/${props.id}`}>Edit</Link>
      </button>
    </div>
  );
};

export default ExpenseListItem;
