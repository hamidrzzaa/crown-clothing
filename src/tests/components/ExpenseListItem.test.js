import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { ExpenseListItem } from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render a single list item", () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
