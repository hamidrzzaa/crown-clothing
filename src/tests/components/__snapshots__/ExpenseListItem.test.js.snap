// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a single list item 1`] = `
<div>
  <h3>
    wolf
  </h3>
  <p>
    1200
  </p>
  <p>
    3
    :
    30
  </p>
  <button>
     
    <[object Object]
      to="/edit/1"
    >
      Edit
    </[object Object]>
  </button>
</div>
`;
